File: lib/models/prise.dart
class Prise {
  final String id;
  final String pecheurId;
  final String maryeurId;
  final String? veterinaireId;
  final String date;
  final String lieu;
  final double? latitude;
  final double? longitude;
  final String? description;
  final String? photo;
  final bool isValid;

  Prise({
    required this.id,
    required this.pecheurId,
    required this.maryeurId,
    this.veterinaireId,
    required this.date,
    required this.lieu,
    this.latitude,
    this.longitude,
    this.description,
    this.photo,
    this.isValid = false,
  });

  factory Prise.fromMap(Map<String, dynamic> map) {
    return Prise(
      id: map['_id']?.toString() ?? map['id']?.toString() ?? '',
      pecheurId:
          map['pecheur']?.toString() ?? map['pecheurId']?.toString() ?? '',
      maryeurId:
          map['maryeur']?.toString() ?? map['maryeurId']?.toString() ?? '',
      veterinaireId:
          map['veterinaire']?.toString() ?? map['veterinaireId']?.toString(),
      date: map['date'] ?? '',
      lieu: map['lieu'] ?? '',
      latitude: _parseDouble(map['latitude']),
      longitude: _parseDouble(map['longitude']),
      description: map['description'],
      photo: map['photo'],
      isValid: map['isValid'] ?? map['isValidated'] ?? false,
    );
  }

  Map<String, dynamic> toMap() {
    final map = {
      'id': id,
      'pecheur': pecheurId,
      'maryeur': maryeurId,
      'date': date,
      'lieu': lieu,
      'latitude': latitude,
      'longitude': longitude,
      'description': description,
      'photo': photo,
      'isValid': isValid,
    };

    // Ajouter le vétérinaire seulement s'il est défini
    if (veterinaireId != null) {
      map['veterinaire'] = veterinaireId;
    }

    return map;
  }

  /// Convertit une valeur en double
  static double? _parseDouble(dynamic value) {
    if (value == null) return null;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) {
      try {
        return double.parse(value);
      } catch (_) {
        return null;
      }
    }
    return null;
  }
}

File: backend/src/models/Prise.js
const mongoose = require('mongoose');

const priseSchema = new mongoose.Schema({
  // Référence au pêcheur qui a effectué la prise
  pecheur: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Pecheur',
    required: true
  },

  // Référence au mareyeur qui gère la prise
  maryeur: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Maryeur',
    required: true
  },

  // Référence au vétérinaire qui validera la prise
  veterinaire: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Veterinaire'
  },

  // Nom de la prise (ex: "Prise du 15/05/2025")
  nom: {
    type: String,
    required: true,
    trim: true
  },

  // Date et heure de début de la pêche
  debut: {
    type: Date,
    required: true
  },

  // Date et heure de fin de la pêche
  fin: {
    type: Date,
    required: true
  },

  // Lieu de la pêche (texte descriptif)
  lieu: {
    type: String,
    trim: true
  },

  // Coordonnées géographiques
  latitude: {
    type: Number
  },
  longitude: {
    type: Number
  },

  // Description supplémentaire de la prise
  description: {
    type: String,
    trim: true
  },

  // URL ou chemin vers une photo de la prise
  photo: String,

  // Informations sur la pêche
  engin: String,  // Type d'engin utilisé
  zone: String,   // Zone de pêche

  // Dates importantes
  affectationDate: {
    type: Date,
    required: true
  },
  dateDebarquement: {
    type: Date,
    required: true
  },

  // Statut de validation
  isValid: {
    type: Boolean,
    default: false
  }
}, {
  timestamps: true
});

// Ajouter des index pour améliorer les performances des requêtes
priseSchema.index({ pecheur: 1 });
priseSchema.index({ maryeur: 1 });
priseSchema.index({ veterinaire: 1 });
priseSchema.index({ debut: -1 });
priseSchema.index({ fin: -1 });
priseSchema.index({ isValid: 1 });

const Prise = mongoose.model('Prise', priseSchema);

module.exports = Prise;